olp-cpp-sdk-core/src/http/DefaultNetwork.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2020 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include <algorithm>
+
+#include "olp/core/http/NetworkConstants.h"
+#include "olp/core/http/NetworkUtils.h"
+
+#include "DefaultNetwork.h"
+
+namespace olp {
+namespace http {
+DefaultNetwork::DefaultNetwork(std::shared_ptr<Network> network)
+    : network_{std::move(network)} {}
+
+SendOutcome DefaultNetwork::Send(NetworkRequest request, Payload payload,
+                                 Callback callback,
+                                 HeaderCallback header_callback,
+                                 DataCallback data_callback) {
+  {
+    auto& request_headers = request.GetMutableHeaders();
+
+    std::unique_lock<std::mutex> lock(default_headers_mutex_);
+    AppendUserAgent(request_headers);
+    AppendDefaultHeaders(request_headers);
+  }
+
+  return network_->Send(std::move(request), std::move(payload),
+                        std::move(callback), std::move(header_callback),
+                        std::move(data_callback));
+}
+
+void DefaultNetwork::Cancel(RequestId id) { network_->Cancel(id); }
+
+void DefaultNetwork::SetDefaultHeaders(Headers headers) {
+  std::unique_lock<std::mutex> lock(default_headers_mutex_);
+  default_headers_ = std::move(headers);
+  user_agent_ = NetworkUtils::ExtractUserAgent(default_headers_);
+}
+
+void DefaultNetwork::AppendUserAgent(Headers& request_headers) const {
+  if (!user_agent_.empty()) {
```

Comment:
I would rather inverse the logic: return on user_agent_.empty().
This way we don't get so nested.